### ISSUE TITLE
[Request] Display key packages' versions along with quasar version

### ISSUE THREAD

**Issue by spectrolite** [quasarframework/quasar-cli#41]
It would be greatly useful (for bug reporting as well as other uses) if quasar version also showed the currently installed versions of at least npm and Vue, and maybe a few other key packages, like quasar-extras or webpack.
```
> quasar version -v
Quasar CLI v0.6.2
webpack  v2.2.1
npm v2.15.11
  
  App running on Quasar v0.14.0 (edge version)
  quasar-extras  v0.0.7
  vue  v2.3.4
```
Just a suggestion. What do you guys think ?

**Comment by JCharante** [comment]
I think that instead of using the `-v` flag, we should use `-a` to log the extra versions.

Example:

```bash
$ quasar version
Quasar CLI v0.6.2
$ quasar version -a
Quasar CLI v0.6.2
webpack v2.2.1
npm v2.15.11
vue v2.3.9
```

**Comment by rstoenescu** [comment]
Sounds like a very good idea. Let's make a list of what other things to include too. Not guaranteeing they will get in, but let's see what other useful we might include.

**Comment by spectrolite** [comment]
that -v for verbose really felt out of place, thanks for a great suggestion.

**Comment by spectrolite** [comment]
I’d say the first “paragraph” of output should be what you would cut and paste into a bug report (quasar cli, running version, npm, webpack, vue.
After those we could add babel, eslint, and maybe allow a user-defined list of packages to track?

**Comment by rstoenescu** [comment]
I think quasar version command should display all info that it can provide without an extra flag, just to ease the usage. I think most users are used to the command without any flag and will continue to use it like this, making us require additional info each time a ticket is opened. We can avoid this by not requiring the flag.

**Comment by spectrolite** [comment]
The flag was just me being humble and not assuming you would adopt my suggestion as default behaviour right from the start. But I can only agree with you.

**Comment by JCharante** [comment]
@rstoenescu I can agree with that. I was just thinking of a scenario where I'd hate having to scroll up to see the Quasar CLI version if there were enough versions logged and keeping it backwards compatible.

**Comment by spectrolite** [comment]
If the list gets very long, repeating the first 4 lines, or moving them to the end, would totally make sense.

**Comment by rstoenescu** [comment]
Makes sense. Along moving Q CLI and Q version being used at end of list, we could include a '-s' flag (--simple) to output just those two versions.

**Comment by spectrolite** [comment]
I insist that instead of just Q-CLI and Q-current,
> first [or last] “paragraph” of output should be what you would cut and paste into a bug report (quasar cli, running version, npm, webpack, vue).

**Comment by JCharante** [comment]
We'll have to remember to edit `.github/ISSUE_TEMPLATE.md` for each repo, replacing

```
Node:
NPM:
```

with

```
<!-- run "quasar version -s" and replace this with the result -->
```

**Comment by spectrolite** [comment]
just a quick thought, what if `quasar version` was the simple one (the one with just what’s useful for a bug report), not very different from now, and we could then use `quasar versions` to get the fully verbose list.

**Comment by spectrolite** [comment]
I think I still prefer Razvan’s point of view of having `quasar version`become much more powerful, and use a flag or maybe another keyword to get the short version.

**Comment by smolinari** [comment]
I like the idea of just making `quasar version` show the major versions of the software used. The list shouldn't be too long and it is all good info for any issue report. If the list does unexpectedly get longer for whatever reason, the `-s` idea from Razvan could be added.  

This first extended list of versions could be something like this?

```
OS Windows 10 Pro
Node v6.10.2
Npm v4.5.0
Quasar v0.14-beta
Quasar CLI v0.6.2
Vue v2.3.9
Webpack v2.2.1
```

Anything missing? Is the order ok? 

Scott

**Comment by smolinari** [comment]
Or, maybe add the `quasar versions` command? So, `quasar version` would only print out the Quasar specific versions and `quasar versions` would print out the extended list????

Hmm...:smile:

Scott

**Comment by JCharante** [comment]
@smolinari Due to the name, I'd think that `quasar versions` would print the version of Quasar & Quasar-CLI, not the versions of everything else.

I personally like the idea of `-a` which logs the library versions in a project, `-q` which logs the quasar versions, and `` which logs the version of quasar cli.

Example:

```bash
~$ quasar version
Quasar CLI v0.6.2

~$ quasar version -a
You\'re not in a Quasar Project Folder

~$ quasar version -q
You\'re not in a Quasar Project Folder

~/QuasarProject$ quasar version -a
OS Ubuntu 17.04
Node v6.10.2
Npm v4.5.0
Quasar v0.14-beta
Quasar CLI v0.6.2
Vue v2.3.9
Webpack v2.2.1

~/QuasarProject$ quasar version -q
Quasar CLI v0.6.2
Quasar 0.14-beta
```

**Comment by spectrolite** [comment]
I’d structure the order of non-quasar versions in a way I can only call “from the outside in”, and I would also add quasar-extras to the mix (and whatever other quasar extension we get in the future), and a few other changes you can deduce from this example, basically a mix of Razvan’s “lets make version way cool” immediately and ideas from both Scott and John.
```
~$ quasar version
Quasar CLI v0.6.2
You're not in a Quasar Project Folder

~/QuasarProject$ quasar version
quasar version -h for options

OS Ubuntu 17.04
Node v6.10.2
Npm v4.5.0
Webpack v2.2.1
Vue v2.3.9
<more here if -a or -c option is used>

Quasar CLI v0.6.2
Quasar Framework v0.14-beta
Quasar Extras v0.0.7

~$ quasar version -q
Quasar CLI v0.6.2
"You're not in a Quasar Project Folder"

~/QuasarProject$ quasar version -q
Quasar CLI v0.6.2
Quasar Framework v0.14.0
Quasar Extras v0.0.7

~$ quasar version -h
Options:
 -a to list all packages (dependencies and devDependencies)
 -c to list all prod packages (dependencies only)
 -q to list quasar versions only
```

**Comment by smolinari** [comment]
Ok. I'd probably also add a separators between the output. Something along the lines of "System", "Dependencies" and "Dev Dependencies". For instance, with your -d example.

```
~/QuasarProject$ quasar version -d
System
  Ubuntu 17.04
  Node 6.10.2
  Npm 4.5.0
  Quasar CLI 0.6.2
  Quasar Framework 0.14.0

Dependencies 
  babel-runtime 6.0.0
  firebase 4.1.2
  firebaseui 2.1.1
  geofire 4.1.2
  gmaps.core 0.5.1
  i18next 8.4.0
  i18next-browser-languagedetector 1.0.1
  vue 2.3.4
  vue-i18next 0.9.0
  vue-params 0.9.2
  vue-router 2.5.3
  vuefire 1.4.3
```

Scott

**Comment by JCharante** [comment]
@smolinari If we're going for that, then we might as well include markup compatible formatting, as a main usecase is for pasting into bug reports.

Example:

```
~/QuasarProject$ quasar version -d
- System
  - Ubuntu 17.04
  - Node 6.10.2
  - Npm 4.5.0
  - Quasar CLI 0.6.2
  - Quasar Framework 0.14.0

- Dependencies 
  - babel-runtime 6.0.0
  - firebase 4.1.2
  - firebaseui 2.1.1
  - geofire 4.1.2
  - gmaps.core 0.5.1
  - i18next 8.4.0
  - i18next-browser-languagedetector 1.0.1
  - vue 2.3.4
  - vue-i18next 0.9.0
  - vue-params 0.9.2
  - vue-router 2.5.3
  - vuefire 1.4.3
```

which will look like this when rendered with markdown

- System
  - Ubuntu 17.04
  - Node 6.10.2
  - Npm 4.5.0
  - Quasar CLI 0.6.2
  - Quasar Framework 0.14.0

- Dependencies 
  - babel-runtime 6.0.0
  - firebase 4.1.2
  - firebaseui 2.1.1
  - geofire 4.1.2
  - gmaps.core 0.5.1
  - i18next 8.4.0
  - i18next-browser-languagedetector 1.0.1
  - vue 2.3.4
  - vue-i18next 0.9.0
  - vue-params 0.9.2
  - vue-router 2.5.3
  - vuefire 1.4.3

**Comment by spectrolite** [comment]
@JCharante anything we can do about the insufficient whitespace here?
![screen shot 2017-07-26 at 19 17 23](https://user-images.githubusercontent.com/1143079/28634303-15e11a8e-7237-11e7-9029-d4a804324887.png)

@smolinari I like the idea of separators/titles. But “System" is inaccurate imho. Some part of me would want to come up with something like “Core stack”, and some part of me says, if it’s the core AND the first section, why does it need a title ?

**Comment by smolinari** [comment]
The top level needs a title, simply because it is expected for a "table of contents" or a "list of things" like that.

What is wrong with "System" more specifically?

Scott

**Comment by JCharante** [comment]
@smolinari What about Root Dependencies

**Comment by JCharante** [comment]
@smolinari I could see the OS under System, but including Node just feels wrong. My system doesn't run on Node.

**Comment by spectrolite** [comment]
@JCharante
> What about Root Dependencies
I meant the vertical space between “Quasar Framework 0.14.0” and “Dependencies”
`Root dependencies` is a new term in an already complex nomenclature, lets keep it simple, and constrained to what most people already know.

**Comment by spectrolite** [comment]
formatting suggestion (which answers my own question btw)
notice double spaces before version number for added readability, and position of blocks (considering this is a long output)


```
~/QuasarProject$ quasar version -a
#### Dev Dependencies
  - autoprefixer  6.5.0
  - babel-core  6.17.0
  - babel-eslint  7.0.0
  - babel-loader  7.0.0
  - babel-plugin-transform-remove-console  6.8.4
  - babel-plugin-transform-runtime  6.15.0
  - babel-preset-es2015  6.16.0
  - babel-preset-es2015-rollup  3.0.0
  - babel-preset-stage-2  6.17.0
  - colors  1.1.2
    …
   <about 30 lines later>
    …
  - vue-devtools  3.1.2
  - vue-loader  12.0.3
  - vue-style-loader  3.0.1
  - vue-template-compiler  2.3.2
  - webpack-dev-middleware  1.8.4
  - webpack-hot-middleware  2.18.0
  - webpack-merge  4.1.0
  - webpack-visualizer-plugin  0.1.11

#### Dependencies 
  - babel-runtime  6.0.0
  - firebase  4.1.2
  - firebaseui  2.1.1
  - geofire  4.1.2
  - gmaps.core  0.5.1
  - i18next  8.4.0
  - i18next-browser-languagedetector  1.0.1
  - vue-i18next  0.9.0
  - vue-params  0.9.2
  - vue-router  2.5.3
  - vuefire  1.4.3

#### Core
  - Ubuntu  17.04
  - Node  6.10.2
  - Npm  4.5.0
  - Webpack  2.2.1
  - Vue  2.3.4
  - Quasar CLI  0.6.2
  - Quasar Framework  0.14.0
  - Quasar Extras  0.0.7
```
#### Dev Dependencies
  - autoprefixer  6.5.0
  - babel-core  6.17.0
  - babel-eslint  7.0.0
  - babel-loader  7.0.0
  - babel-plugin-transform-remove-console  6.8.4
  - babel-plugin-transform-runtime  6.15.0
  - babel-preset-es2015  6.16.0
  - babel-preset-es2015-rollup  3.0.0
  - babel-preset-stage-2  6.17.0
  - colors  1.1.2
    …
   <about 30 lines later>
    …
  - vue-devtools  3.1.2
  - vue-loader  12.0.3
  - vue-style-loader  3.0.1
  - vue-template-compiler  2.3.2
  - webpack-dev-middleware  1.8.4
  - webpack-hot-middleware  2.18.0
  - webpack-merge  4.1.0
  - webpack-visualizer-plugin  0.1.11

#### Dependencies 
  - babel-runtime  6.0.0
  - firebase  4.1.2
  - firebaseui  2.1.1
  - geofire  4.1.2
  - gmaps.core  0.5.1
  - i18next  8.4.0
  - i18next-browser-languagedetector  1.0.1
  - vue-i18next  0.9.0
  - vue-params  0.9.2
  - vue-router  2.5.3
  - vuefire  1.4.3

#### Core
  - Ubuntu  17.04
  - Node  6.10.2
  - Npm  4.5.0
  - Webpack  2.2.1
  - Vue  2.3.4
  - Quasar CLI  0.6.2
  - Quasar Framework  0.14.0
  - Quasar Extras  0.0.7

**Comment by JCharante** [comment]
@spectrolite That makes it look much nicer. :+1:

**Comment by spectrolite** [comment]
I’m done fiddling now, I think it’s starting to look quite nice too.

**Comment by spectrolite** [comment]
```
~/QuasarProject$ quasar version
quasar version -h for options

###  MyAppName  0.7.2

    *  Ubuntu  17.04
    *  Node  6.10.2
    *  Npm  4.5.0
    *  Webpack  2.2.1
    *  Vue  2.3.4

    *  Quasar CLI  0.6.2
    *  Quasar Framework  0.14.0
    *  Quasar Extras  0.0.7
```

###  MyAppName  0.7.2

    *  Ubuntu  17.04
    *  Node  6.10.2
    *  Npm  4.5.0
    *  Webpack  2.2.1
    *  Vue  2.3.4

    *  Quasar CLI  0.6.2
    *  Quasar Framework  0.14.0
    *  Quasar Extras  0.0.7

**Comment by JCharante** [comment]
@spectrolite I have to disagree with the latest mockup. It's silly to include the project name & version, because most of the time it's just going to be quasar-app 0.0.1

**Comment by spectrolite** [comment]
@JCharante Well, it shouldn’t. It’s bad practice. And if having ###quasar-app 0.0.1 slapped in your face each time you query for the version helps you finally start proper versioning and naming, I think it’s a good thing. ;-)

**Comment by spectrolite** [comment]
@JCharante if you don’t care about it, you won’t mind having it there, and its position on top gives it the lesser importance of all (remember we’re in command line).

**Comment by JCharante** [comment]
@spectrolite Well I was getting at how a lot of bug reports I've seen seem to be from people who have recently started a project that encounters a "bug". I don't usually start using semver until there's an MVP and it's being used / deployed on something other than my dev computer. I'm not going to die over seeing

### quasar-project 0.0.1

when someone includes it in a bug report or when I check my quasar version, but I just don't see the value added by it.

Actually, now I've seen the light. The project name is the root and the other sections are its branches (with the individual packages being leafs). I love it. :+1:

**Comment by spectrolite** [comment]
updated last 2 mockups with your suggestion of H2.

**Comment by spectrolite** [comment]
tbh, I don’t see MUCH value in it either.
It just seemed weird that we now have a tool that displays not only quasar’s internals’ versions, but also all packages, and the OS, and node, and npm… basically everything in package.json and then some, except for the actual app version. It really felt like we were ostracising that piece of VERSION info for no good reason. I agree it’s often going to be irrelevant. But it has to be there, at least for the sake of coherence.

**Comment by JCharante** [comment]
Any thoughts @rstoenescu?

**Comment by rstoenescu** [comment]
I'm seeing this implemented as:

$ quasar version

##   MyAppName  0.7.2
  - Ubuntu 17.04
  - Node 6.10.2
  - Npm 4.5.0
  - Webpack 2.2.1
  - Vue 2.3.4
  - Quasar CLI 0.6.2
  - Quasar Framework 0.14.0
  - Quasar Extras 0.0.7

$ quasar version -q

##   MyAppName  0.7.2
  - Vue 2.3.4
  - Quasar Framework 0.14.0
  - Quasar Extras 0.0.7

**Comment by spectrolite** [comment]
that looks perfect.
what about ` quasar version -a ` like this also ?
 https://github.com/quasarframework/quasar-cli/issues/41#issuecomment-318151812

**Comment by rstoenescu** [comment]
@spectrolite Yes, let's add that too.
@smolinari I think we can go ahead with this. Shall I assign it to you?

**Comment by smolinari** [comment]
@rstoenescu - Sure. But, don't expect speedy results. 

Scott

**Comment by rstoenescu** [comment]
This has been implemented in Quasar CLI 0.15.0-beta, so closing this.
